staging/src/k8s.io/apiserver/plugin/pkg/authenticator/token/oidc/oidc.go

```diff
@@ -238,43 +325,272 @@ func newAuthenticator(opts Options, initVerifier func(ctx context.Context, a *Au
 		Now:                  now,
 	}
 
+	authenticator := &Authenticator{
+		issuerURL:       opts.IssuerURL,
+		usernameClaim:   opts.UsernameClaim,
+		usernamePrefix:  opts.UsernamePrefix,
+		groupsClaim:     opts.GroupsClaim,
+		groupsPrefix:    opts.GroupsPrefix,
+		requiredClaims:  opts.RequiredClaims,
+		cancel:          cancel,
+		claimsProcessor: newClaimResolver(ctx, opts.IssuersPerClaim, tr, verifierConfig),
+	}
+
 	initVerifier(ctx, authenticator, verifierConfig)
 	return authenticator, nil
 }
 
-func hasCorrectIssuer(iss, tokenData string) bool {
-	parts := strings.Split(tokenData, ".")
+// untrustedIssuer extracts an untrusted "iss" claim from the given JWT token,
+// or returns an error if the token can not be parsed.  Since the JWT is not
+// verified, the returned issuer should not be trusted.
+func untrustedIssuer(token string) (string, error) {
+	parts := strings.Split(token, ".")
 	if len(parts) != 3 {
-		return false
+		return "", fmt.Errorf("malformed token %q", token)
 	}
 	payload, err := base64.RawURLEncoding.DecodeString(parts[1])
 	if err != nil {
-		return false
+		return "", fmt.Errorf("error decoding token %q", token)
 	}
 	claims := struct {
 		// WARNING: this JWT is not verified. Do not trust these claims.
 		Issuer string `json:"iss"`
 	}{}
 	if err := json.Unmarshal(payload, &claims); err != nil {
+		return "", fmt.Errorf("while unmarshaling token: %q: %v", err)
+	}
+	return claims.Issuer, nil
+}
+
+func hasCorrectIssuer(iss, tokenData string) bool {
+	uiss, err := untrustedIssuer(tokenData)
+	if err != nil {
+		glog.Errorf("while checking issuer: %v", err)
 		return false
 	}
-	if claims.Issuer != iss {
+	if uiss != iss {
+		glog.V(11).Infof("Issuer mismatch: want=%q, got=%q", iss, uiss)
```

Comment:
`V(1)` instead of 11 and 10 throughout this file?

Reply:
I'd rather not: these bits of information are only useful if debugging.

AFAIK, K8S apiserver has --v=2 set for many deployments.  Moving this to V(1) would make log outputs waaaay too noisy, IMHO.  What do you think?

Reply:
Then how about 3?

Reply:
There's a bunch of logging conventions in Kubernetes that seem to be relevant for this decision. I was vaguely aware of them when writing this code, but your comments made me go back and check in detail.  Here's what I found.

For one, there is https://github.com/kubernetes/community/blob/master/contributors/devel/logging.md, which pretty much means level 1..4 are taken.

Also AFAIK `kubectl` uses log levels up to 8 for various levels of detail about HTTPS requests.  
See: https://github.com/ianchakeres/kubernetes.github.io/blob/51d19f271c5f5217d5e68418c997a5bd4de77f20/docs/user-guide/kubectl-cheatsheet.md#kubectl-output-verbosity-and-debugging This means we can normally expect end users to run up to `kubectl --v=8`.  So I'd rather not pollute those levels with very low level information.

Since the log messages here are not intended to be surfaced in the logs *unless* we specifically want to hunt for therm (that is, you'd use `--vmodule`, and not `--v` to turn them on), it seems to follow that the log level may only be something greater than 8.  But then, why not pick a nice round number as 10 as the baseline?  

Hence 10, 11 seem to make sense to me.  Requiring to change that to 9, 10 doesn't seem like a big win.

Reply:
running a server at 6 or up isn't generally reasonable... that starts to log all requests made by internal transports at various verbosity levels. I'd generally expect trace debugging like this around level 4 or 5

Reply:
Moved to levels 4 and 5.